### DISCUSSION TITLE
Now available in PhotoDemon: Content-Aware Fill (aka Heal Selection)

### DISCUSSION THREAD

**Discussion by tannerhelland** [tannerhelland/PhotoDemon#402] [News]
![content-aware fill](https://user-images.githubusercontent.com/1930029/166238705-c437f6f9-bed7-4707-9c4e-5e43b121910f.png)

This tool is in its final development stages.  I'm currently building a user interface for it (to display a number of optional toggles) and I'm continuing to improve its performance, but the core algorithm is in-place and working well.

This feature was first added to Photoshop in CS5.  The algorithm I'm using for PhotoDemon was first proposed by Paul Harrison in his PhD thesis, "Texture Synthesis, Texture Transfer, and Plausible Restoration":

https://logarithmic.net/pfh/thesis

Paul's work would go on to form the foundation of GIMP's "Resynthesis" plugin.  PhotoDemon's tool will work similarly to GIMP's, but was built from the ground-up for PD.  It requires no external plugins or 3rd-party libraries and is built directly into the app.  Like nearly all PhotoDemon features, it will work on all Windows versions (all the way back to Windows XP) and it has no special hardware requirements.

I expect to merge the final version in the next week or two.  Stay tuned!  😄

**Comment by wqweto** [comment]
![Content-Aware Magic](https://c.tenor.com/wn2_Qq6flogAAAAC/magical-magic.gif)

**Comment by tannerhelland** [comment]
This feature is now available in nightly builds!  Access it by creating a selection and using the `Edit > Content-aware fill` menu, or the `Select > Heal selected area` menu.

**Reply by hi5** [reply]
Nice - sure beats using the cloning tool in PSP

**Comment by g-busch** [comment]
Great tool; I'm having lots of fun making objects vanish from pictures! :) In many cases I can use the tool's default settings.

I have one suggestion: Since there's a cloning brush in the toolbox, why not add the fill item there as a companion to it, since the two are closely related?

**Reply by tannerhelland** [reply]
Great suggestion, @g-busch .  Other software provides a tool like this as a "healing brush", and I would like to do the same.  The primary challenge is that this brush would behave a little differently owing to the performance constraints it has - so the user would need to "paint" a region, and then when the mouse is released, the algorithm would work its magic and fill in the new area all at once.  (The algorithm is not fast enough to paint as-you-go, I fear.)

This requires a new brush framework which is why I haven't added it yet, but a healing brush is on my TODO list.  I'll see if I can't bump it up the priority queue a bit now that the underlying algorithm is largely finished.

**Reply by g-busch** [reply]
I see what you mean. I've found this fill function a very good substitute for the healing brush, since the source in my case is usually in the area surrounding the selection I want to heal. A healing brush, of course, allows selecting a source that's not necessarily in the region of the target, like in the case of the clone brush.

A comparison I was able to make was with the healing brush in FastStone Image Viewer. There I have to first select my source, which I then paint over the target. It works, but is somehow not as effective for what I'm doing as your context-aware fill feature.

At any rate, other users are likely to work differently, so the option of using either of the features will only make PhotoDemon even better than it already is.
